gdnative-sys/src/lib.rs

```diff
@@ -7,46 +7,91 @@
+            want,
+            got,
+        } => {
+            if let Some(f) = (*options).report_version_mismatch {
+                f(
+                    (*options).gd_native_library,
+                    CString::new(format!("{}", api_type)).unwrap().as_ptr(),
+                    want,
+                    got,
+                );
+            }
+        }
+        InitError::NotFound { message } => {
+            if let Some(f) = (*options).report_loading_error {
+                f(
+                    (*options).gd_native_library,
+                    CString::new(message).unwrap().as_ptr(),
+                );
+            }
+        }
+    }
+}
+
 unsafe fn find_api_ptr(
     core_api: *const godot_gdnative_core_api_struct,
     api_type: GDNATIVE_API_TYPES,
     version_major: u32,
     version_minor: u32,
-) -> *const godot_gdnative_api_struct {
-    let mut api = core_api as *const godot_gdnative_api_struct;
-    if (*api).type_ as u32 == api_type as u32 {
-        while !api.is_null() {
-            if (*api).version.minor == version_minor
-                && (*api).version.major == version_major
-                // The boolean expression below SHOULD always be true;
-                // we will double check to be safe.
-                && (*api).type_ as u32 == api_type as u32
-            {
-                return api;
-            }
-            api = (*api).next;
-        }
-    }
-    for i in 0..(*core_api).num_extensions {
-        let mut extension =
-            *(*core_api).extensions.offset(i as _) as *const godot_gdnative_api_struct;
-        if (*extension).type_ as u32 == api_type as u32 {
-            while !extension.is_null() {
-                if (*extension).version.minor == version_minor
-                    && (*extension).version.major == version_major
+) -> Result<*const godot_gdnative_api_struct, InitError> {
+    let mut got = None;
+    let mut find_version =
+        |mut api: *const godot_gdnative_api_struct| -> Option<*const godot_gdnative_api_struct> {
+            if (*api).type_ as u32 == api_type as u32 {
+                while !api.is_null() {
                     // The boolean expression below SHOULD always be true;
                     // we will double check to be safe.
-                    && (*extension).type_ as u32 == api_type as u32
-                {
-                    return extension;
+                    if (*api).type_ as u32 == api_type as u32 {
+                        let (major, minor) = ((*api).version.major, (*api).version.minor);
+                        if major == version_major && minor == version_minor {
+                            return Some(api);
+                        } else {
+                            got = Some(godot_gdnative_api_version { major, minor });
+                        }
+                    }
+                    api = (*api).next;
                 }
-                extension = (*extension).next;
             }
+            None
+        };
```

Comment:
Why the closure and mutable state? This is fragile and non-idiomatic in Rust. `find_version` should just be a separate function that returns `Result<*const godot_gdnative_api_struct, InitError>`.

Reply:
Moved to a separate `find_version` function. However I don't see how it can return `Result<*const godot_gdnative_api_struct, InitError>`; we don't know whether the API is present until we've inspected both `core_api` and all of `core_api.extensions`. Could you elaborate?

btw do you have a reference link for the fragility of closures with mutable state? I'm new to rust and would like to understand the best practices.

Reply:
You may return an `Option<Result>>`, then. Instead of mutating a `&mut` of `got` passed from outside, you may keep a `mut got` binding inside, and return a `Some(Err(InitError::VersionMismatch))` when you can't find a version of a struct, or `None` if the type mismatch. In `find_api_ptr`, you can then keep track of a `last_error: Option<InitError>` instead of `got`, making it easier for a reader to see the intention.

- - -

> reference link for the fragility of closures with mutable state

I don't have a link to a Rust-specific article ready. It's not a problem specific to Rust or closures, but rather a general observation of implicit shared mutable state: it makes programs harder to reason about, and thus harder to maintain.

Implicity is good on itself, but bad when combined with shared mutable state.

When you think about it, mutable closure captures when used this way are a little like globals: you can't see what may be modified at the call site (implicit), and they may modify the same variables as you do (shared and mutable). Each call may or may not invalidate all your assumptions, and you can only know which way by carefully scrutinizing the code for side effects -- and not making any mistakes.

We are human, and we make mistakes. Thus, implicit shared mutable state are undesirable.

Reply:
Done. Thanks for the detailed response!